ironfish/src/wallet/account.ts

```diff
@@ -213,6 +216,156 @@ export class Account {
+    for (const { assetId, value } of transaction.burns) {
+      const existingAsset = await this.walletDb.getAsset(this, assetId, tx)
+      if (!existingAsset) {
+        continue
+      }
+      // Verify the owner matches before processing a burn
+      Assert.isEqual(
+        existingAsset.owner.toString('hex'),
+        this.publicAddress,
+        'Existing asset owner should match public address',
+      )
+
+      const existingSupply = existingAsset.supply
+      const supply = existingSupply - value
+      Assert.isTrue(supply >= BigInt(0), 'Invalid burn value')
+
+      await this.walletDb.putAsset(
+        this,
+        assetId,
+        {
+          createdTransactionHash: existingAsset.createdTransactionHash,
+          id: existingAsset.id,
+          metadata: existingAsset.metadata,
+          name: existingAsset.name,
+          owner: existingAsset.owner,
+          supply,
+        },
+        tx,
+      )
+    }
+  }
+
+  private async deleteDisconnectedBurnsFromAssetsStore(
+    transaction: Transaction,
+    tx: IDatabaseTransaction,
+  ): Promise<void> {
+    for (const { assetId, value } of transaction.burns.slice().reverse()) {
+      const existingAsset = await this.walletDb.getAsset(this, assetId, tx)
+      if (!existingAsset) {
+        continue
+      }
+      // Verify the owner matches before processing a burn
+      Assert.isEqual(
+        existingAsset.owner.toString('hex'),
+        this.publicAddress,
+        'Existing asset owner should match public address',
+      )
+
+      const existingSupply = existingAsset.supply
+      const supply = existingSupply + value
+
+      await this.walletDb.putAsset(
+        this,
+        assetId,
+        {
+          createdTransactionHash: existingAsset.createdTransactionHash,
+          id: existingAsset.id,
+          metadata: existingAsset.metadata,
+          name: existingAsset.name,
+          owner: existingAsset.owner,
+          supply,
+        },
+        tx,
+      )
+    }
+  }
+
+  private async deleteDisconnectedMintsFromAssetsStore(
+    transaction: Transaction,
+    tx: IDatabaseTransaction,
+  ): Promise<void> {
+    for (const { asset, value } of transaction.mints.slice().reverse()) {
+      // Only update the mint for the owner
+      if (asset.owner().toString('hex') !== this.publicAddress) {
+        continue
+      }
+
+      const assetId = asset.id()
+      const existingAsset = await this.walletDb.getAsset(this, assetId, tx)
+      Assert.isNotUndefined(existingAsset)
+
+      const existingSupply = existingAsset.supply
+      const supply = existingSupply - value
+      Assert.isTrue(supply >= BigInt(0))
+
+      // If we are reverting the transaction which matches the created at
+      // hash of the asset, delete the record from the store
+      if (
+        transaction.hash().equals(existingAsset.createdTransactionHash) &&
+        supply === BigInt(0)
+      ) {
+        await this.walletDb.deleteAsset(this, assetId, tx)
```

Comment:
You might be planning to address this with assets in pending transactions -- the transaction may be pending after it's disconnected so we may not want to delete the asset.

Reply:
Yeah, that's a good point. I'll refactor this so the asset gets deleted when the associated created transaction is. Otherwise, we can just adjust the supply.

I'll also add a blockHash / sequence since that may be useful to infer pending status.